Nos/Service/RelayService.swift

```diff
@@ -9,260 +9,258 @@ import Starscream
     
     init(persistenceController: PersistenceController) {
         self.persistenceController = persistenceController
         self.backgroundContext = persistenceController.newBackgroundContext()
+        self.subscriptions = RelaySubscriptionManager()
 
-        self.publishFailedEventsTimer = AsyncTimer(timeInterval: 120, onFire: { [weak self] in
+        self.saveEventsTimer = AsyncTimer(timeInterval: 1, priority: .high) { [weak self] in
+            await self?.backgroundContext.perform(schedule: .immediate) {
+                if self?.backgroundContext.hasChanges == true {
+                    try! self?.backgroundContext.save()
+                }
+            }
+        }
+        
+        self.updateSocialGraphTimer = AsyncTimer(timeInterval: 30, onFire: { @MainActor [weak self] in
+            guard let currentUser = CurrentUser.shared.author else {
+                return
+            }
+            // Close sockets for anything not in the above
+            if let keptRelays = currentUser.relays as? Set<Relay> {
+                await self?.closeAllConnections(excluding: keptRelays)
+            }
+            
+            await CurrentUser.shared.updateInNetworkAuthors()
+        })
+        
+        // TODO: fire this after all relays have connected, not right on init
+        self.publishFailedEventsTimer = AsyncTimer(timeInterval: 60, onFire: { [weak self] in
             await self?.publishFailedEvents()
         })
         
         Task { @MainActor in
             CurrentUser.shared.viewContext = persistenceController.container.viewContext
             await openSockets()
         }
+        
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(appWillEnterForeground),
+            name: UIApplication.willEnterForegroundNotification,
+            object: nil
+        )
     }
     
     deinit {
-        publishFailedEventsTimer?.cancel()
+        saveEventsTimer?.cancel()
+        updateSocialGraphTimer?.cancel()
     }
     
-    // TODO: lock requestQueueLock before calling this
-    var activeSubscriptions: [String] {
-        requestFilterQueue
-            .filter { $0.subscriptionStartDate != nil }
-            .map { $0.subscriptionId }
+    @objc func appWillEnterForeground() {
+        Task { await openSockets() }
     }
-        
-    private func removeFilter(for subscription: String) {
-        requestQueueLock.withLock {
-            if let foundFilterIndex = requestFilterQueue.firstIndex(where: { $0.subscriptionId == subscription }) {
-                requestFilterQueue.remove(at: foundFilterIndex)
-            }
-        }
-    }
-        
-    private func handleError(_ error: Error?) {
+    
+    private func handleError(_ error: Error?, from socket: WebSocketClient) {
         if let error {
-            print("websocket error: \(error)")
+            Log.info("websocket error: \(error) from: \(socket.host)")
```

Comment:
Is INFO the correct level here? I would guess it should be error, but maybe receiving errors from websockets is expectable, so feel free to chose the leven you see fit

Reply:
It happens often enough that I did intentionally go with INFO. I try to reserve ERROR for things that are more actionable for us as developers.